### PR TITLE
System test for cluster role split for cluster wide operator with lim…

### PR DESCRIPTION
**Pull request by see-quick**
…ited access

Signed-off-by: see-quick <maros.orsak159@gmail.com>

### Type of change

- Enhancement / new feature
- Refactoring

### Description

This PR aims **(a)** cover https://github.com/strimzi/strimzi-kafka-operator/pull/7390. I have added one test case, which in summary does the following steps:

1. Define our own Roles, ClusterRoles, RoleBindings and ClusterRoleBindings
2. Deploy cluster-wide Cluster Operator
3. During deployment of ClusterOperator, create ClusterRole (-watched) bound to a cluster-wide configuration
4. During deployment of ClusterOperator, create ClusterRole (-namespaced) bound to some specific namespace (i.e., `example-1`)
5. Deploy Kafka cluster into example-namespace
6. Deploy Kafka cluster into infra-namespace
7. Check that Kafka CR is present in infra-namespace but fails (it can be checked via CR Status)
8. Check that Kafka CR is present and successfully deployed in example-namespace 

(b) adding functionality to our SetupClusterOperator (i.e., a user can now add his own Role, RoleBinding, ClusterRole and ClusterRoleBinding resources and pass it via SetupClusterOperator builder. 

### Checklist

- [x] Write tests
- [x] Make sure all tests pass

### CODE REVIEW

File: systemtest/src/test/java/io/strimzi/systemtest/specific/SpecificIsolatedST.java [hunk 1]
```diff
@@ -193,6 +207,97 @@ void testLoadBalancerSourceRanges(ExtensionContext extensionContext) {
...
+
+        final List<Role> roles = Arrays.asList(strimziClusterOperator020);
+        final List<ClusterRole> clusterRoles = Arrays.asList(strimziClusterOperator021, strimziClusterOperator022,
+                strimziClusterOperator023, strimziClusterOperator030, strimziClusterOperator031, strimziClusterOperator033);
+
+        // ---- b) defining RoleBindings
+        final RoleBinding strimziClusterOperator020Namespaced = TestUtils.configFromYaml(SetupClusterOperator.getInstanceHolder().switchClusterRolesToRolesIfNeeded(new File(Constants.PATH_TO_PACKAGING_INSTALL_FILES + "/cluster-operator/020-RoleBinding-strimzi-cluster-operator.yaml"), true), RoleBinding.class);
+        final RoleBinding strimziClusterOperator022LeaderElection = TestUtils.configFromYaml(SetupClusterOperator.getInstanceHolder().changeLeaseNameInResourceIfNeeded(new File(Constants.PATH_TO_PACKAGING_INSTALL_FILES + "/cluster-operator/022-RoleBinding-strimzi-cluster-operator.yaml").getAbsolutePath()), RoleBinding.class);
```
Comment (im-konge) [inline systemtest/src/test/java/io/strimzi/systemtest/specific/SpecificIsolatedST.java:235]:
```suggestion
        final RoleBinding strimziClusterOperator022LeaderElection = TestUtils.configFromYaml(SetupClusterOperator.getInstanceHolder().changeLeaseNameInResourceIfNeeded(new File(Constants.PATH_TO_LEASE_ROLE_BINDING).getAbsolutePath()), RoleBinding.class);
```

File: systemtest/src/test/java/io/strimzi/systemtest/specific/SpecificIsolatedST.java [hunk 2]
```diff
@@ -193,6 +207,97 @@ void testLoadBalancerSourceRanges(ExtensionContext extensionContext) {
...
+            .runBundleInstallation();
+
+        // implicit verification that a user is able to deploy Kafka cluster in namespace <example-1>, where we are allowed
+        // to create Custom Resources because of `*-namespaced Role`
+        resourceManager.createResource(extensionContext, false, KafkaTemplates.kafkaEphemeral(testStorage.getClusterName(), 3)
+            .editMetadata()
+            // this should work
+            .withNamespace(namespaceWhereCreationOfCustomResourcesIsApproved)
+            .endMetadata()
+            .build());
```
Comment (im-konge) [inline systemtest/src/test/java/io/strimzi/systemtest/specific/SpecificIsolatedST.java:279]: Indent
Also, if it should work, shouldn't we wait for readiness of Kafka?
Or deploy it in parallel and wait for first Kafka to be ready and second Kafka to contain message about failure?

File: systemtest/src/test/java/io/strimzi/systemtest/specific/SpecificIsolatedST.java [hunk 3]
```diff
@@ -193,6 +207,97 @@ void testLoadBalancerSourceRanges(ExtensionContext extensionContext) {
...
+
+        resourceManager.createResource(extensionContext, false, KafkaTemplates.kafkaEphemeral(testStorage.getClusterName(), 3)
+            .editMetadata()
+                // this should not work
+                .withNamespace(clusterOperator.getDeploymentNamespace())
+            .endMetadata()
+            .build());
+
+        // verify that in `infra-namespace` we are not able to deploy Kafka cluster
+        KafkaUtils.waitForKafkaStatusUpdate(clusterOperator.getDeploymentNamespace(), testStorage.getClusterName());
```
Comment (im-konge) [inline systemtest/src/test/java/io/strimzi/systemtest/specific/SpecificIsolatedST.java:289]: Would it be more stable if we go with `waitUntilKafkaStatusConditionContainsMessage()`?
I'm afraid that there can be few moments when we don't get desired failure message

File: systemtest/src/test/java/io/strimzi/systemtest/specific/SpecificIsolatedST.java [hunk 4]
```diff
@@ -193,6 +207,97 @@ void testLoadBalancerSourceRanges(ExtensionContext extensionContext) {
...
+        final ClusterRole strimziClusterOperator022 = TestUtils.configFromYaml(SetupClusterOperator.getInstanceHolder().changeLeaseNameInResourceIfNeeded(new File(Constants.PATH_TO_PACKAGING_INSTALL_FILES + "/cluster-operator/022-ClusterRole-strimzi-cluster-operator-role.yaml").getAbsolutePath()), ClusterRole.class);
+        final ClusterRole strimziClusterOperator023 = TestUtils.configFromYaml(new File(Constants.PATH_TO_PACKAGING_INSTALL_FILES + "/cluster-operator/023-ClusterRole-strimzi-cluster-operator-role.yaml"), ClusterRole.class);
+        final ClusterRole strimziClusterOperator030 = TestUtils.configFromYaml(new File(Constants.PATH_TO_PACKAGING_INSTALL_FILES + "/cluster-operator/030-ClusterRole-strimzi-kafka-broker.yaml"), ClusterRole.class);
+        final ClusterRole strimziClusterOperator031 = TestUtils.configFromYaml(new File(Constants.PATH_TO_PACKAGING_INSTALL_FILES + "/cluster-operator/031-ClusterRole-strimzi-entity-operator.yaml"), ClusterRole.class);
+        final ClusterRole strimziClusterOperator033 = TestUtils.configFromYaml(new File(Constants.PATH_TO_PACKAGING_INSTALL_FILES + "/cluster-operator/033-ClusterRole-strimzi-kafka-client.yaml"), ClusterRole.class);
```
[the remark this answers is not in the document]
Comment (im-konge) [reply systemtest/src/test/java/io/strimzi/systemtest/specific/SpecificIsolatedST.java:227]: Maybe having this kind of method also for `RoleBinding`s would make also sense - then we can remove from list what we don't need

**Review by im-konge — COMMENTED** [review]
Few comments, but this looks really good :) Thank you!

File: systemtest/src/main/java/io/strimzi/systemtest/resources/operator/SetupClusterOperator.java [hunk 5]
```diff
@@ -253,7 +264,19 @@ private void bundleInstallation() {
         // check if namespace is already created
         createCONamespaceIfNeeded();
         prepareEnvForOperator(extensionContext, namespaceInstallTo, bindingsNamespaces);
-        applyBindings();
+        // if a user want to customise one of the Role, ClusterRole, RoleBindings and ClusterRoleBinding he must do it
+        // everything by himself in scope of RBAC permissions otherwise we apply the default one
```
Comment (scholzj) [inline systemtest/src/main/java/io/strimzi/systemtest/resources/operator/SetupClusterOperator.java:268]: This is not something a _user_ would do in system tests. Should it instead say _if they are managed directly by the individual test_?

File: systemtest/src/test/java/io/strimzi/systemtest/specific/SpecificIsolatedST.java [hunk 6]
```diff
@@ -193,6 +207,97 @@ void testLoadBalancerSourceRanges(ExtensionContext extensionContext) {
...
+        KafkaUtils.waitForKafkaStatusUpdate(clusterOperator.getDeploymentNamespace(), testStorage.getClusterName());
+
+        final Condition condition = KafkaResource.kafkaClient().inNamespace(clusterOperator.getDeploymentNamespace()).withName(testStorage.getClusterName()).get().getStatus().getConditions().stream().findFirst().get();
+
+        assertThat(condition.getReason(), CoreMatchers.is("KubernetesClientException"));
+        assertThat(condition.getStatus(), CoreMatchers.is("True"));
+        assertThat(condition.getMessage(), CoreMatchers.containsString("Forbidden!Configured service account doesn't have access"));
+
+        // rollback to the default configuration
+        clusterOperator.rollbackToDefaultConfiguration();
```
Comment (scholzj) [inline systemtest/src/test/java/io/strimzi/systemtest/specific/SpecificIsolatedST.java:298]: What exactly does this mean?

**Review by scholzj — COMMENTED** [review]
I do not see anywhere the deletion of the RBAC resource you create in the test. Are they deleted somewhere automatically?

File: systemtest/src/test/java/io/strimzi/systemtest/specific/SpecificIsolatedST.java (hunk 2 above)
Comment (see-quick) [reply systemtest/src/test/java/io/strimzi/systemtest/specific/SpecificIsolatedST.java:279]: > Also, if it should work, shouldn't we wait for readiness of Kafka?

Yes, that's leftover...

> Or deploy it in parallel and wait for first Kafka to be ready and second Kafka to contain message about failure?

Well, I will re-order the deployment of these Kafka's (and we can achieve that kind of parallelism with additional readiness check).

File: systemtest/src/test/java/io/strimzi/systemtest/specific/SpecificIsolatedST.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (see-quick) [reply systemtest/src/test/java/io/strimzi/systemtest/specific/SpecificIsolatedST.java:227]: Well, in this case, I want explicitly to have ClusterRole and not Role (as we implemented in the SetupClusterOperator).

File: systemtest/src/main/java/io/strimzi/systemtest/resources/operator/SetupClusterOperator.java (hunk 5 above)
Comment (see-quick) [reply systemtest/src/main/java/io/strimzi/systemtest/resources/operator/SetupClusterOperator.java:268]: Yes, that's better :) I will update it.

File: systemtest/src/test/java/io/strimzi/systemtest/specific/SpecificIsolatedST.java (hunk 6 above)
Comment (see-quick) [reply systemtest/src/test/java/io/strimzi/systemtest/specific/SpecificIsolatedST.java:298]: Well, this actually un-install the Cluster Operator with the specific configuration, which we used in the test case, and then deploys the default Cluster Operator.

**Comment by see-quick** [comment]
> I do not see anywhere the deletion of the RBAC resource you create in the test. Are they deleted somewhere automatically?

That's handled by `clusterOperator.rollbackToDefaultConfiguration();`. It completely uninstalls the current installation of SetupClusterOperator (i.e., custom Roles...) and deploys the one with the default configuration

File: systemtest/src/test/java/io/strimzi/systemtest/specific/SpecificIsolatedST.java (hunk 6 above)
Comment (scholzj) [reply systemtest/src/test/java/io/strimzi/systemtest/specific/SpecificIsolatedST.java:298]: Maybe the comment can be more explicit about it then? How does it know what to uninstall if you install the RBAC bindings in your own way? How do you make sure they are not left behind?

Comment (see-quick) [reply systemtest/src/test/java/io/strimzi/systemtest/specific/SpecificIsolatedST.java:298]: > Maybe the comment can be more explicit about it then? 

Yes, I will more elaborate on this comment :) thanks

> How does it know what to uninstall if you install the RBAC bindings in your own way? How do you make sure they are not left behind?

TL;DR; It is handled by ResourceManager inside SetupClusterOperator.
--
When you create a resource by ResourceManager using the generic method `resourceManager.createResource(extensionContent, <resource>)` it will do the following:
1. Find resource type; meaning if such resource is Kafka, KMM2 etc.
2. Create a resource 
3. and then we store info to the `Map<String, Stack<ResourceItem>>`, the key identifier is the test case name and the Stack contains multiple resources, which will be deleted after the test case because we call `ResourceManager.getInstance().deleteResources(extensionContext);` method, which retrieve all of the resources specific to test case thanks to extensionContent metadata.

In this scenario `clusterOperator.rollbackToDefaultConfiguration();` calls at first unInstall() method, which completely deletes all resources (this is handled by ResourceManager) and all namespaces. Thus also removing RBAC bindings deployed in this test case.

Comment (scholzj) [reply systemtest/src/test/java/io/strimzi/systemtest/specific/SpecificIsolatedST.java:298]: Ok, thanks for the explanation. 👍

**Review by scholzj — APPROVED** [review]

**Review by Frawless — APPROVED** [review]

**Review by im-konge — APPROVED** [review]

**Comment by scholzj** [comment]
@see-quick Does this also work with Helm / OLM etc.?
